### PR TITLE
Parameter object and adaptation

### PR DESCRIPTION
**Pull request by LeahPrice**
This commit is the start of making it easier to perform adaptation in SMC.

The main changes here are:
- Adding a template parameter for the algorithm parameters to the sampler object
- Creating a base class for adaptation
- Doing the MCMC repeats within the library

### CODE REVIEW

**Comment by LeahPrice** [comment]
I changed the MCMC step code a little so that the loop over the repeats is done within the library and users write code for a single MCMC step for a single particle. This seemed helpful for adapting the number of MCMC repeats (it is easier to work out the acceptance probability from previous steps if you number the number of repeats).

I think in a future pull request I’ll store more of the calculations for data annealing SMC. I’m essentially doubling up on the likelihood and prior calculations because I’m recalculating them for the current particle at every iteration.

File: R/RcppExports.R [hunk 1]
```diff
@@ -2,26 +2,26 @@
 # Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393
 
 blockpfGaussianOpt_impl <- function(data, part, lag) {
-    .Call(`_RcppSMC_blockpfGaussianOpt_impl`, data, part, lag)
+    .Call('RcppSMC_blockpfGaussianOpt_impl', PACKAGE = 'RcppSMC', data, part, lag)
```
Comment (eddelbuettel) [inline R/RcppExports.R]: You are using an Rcpp version older than 0.12.12.  

Please upgrade, and re-create the file, or explain why you must use on old version, possibly over at the Rcpp repo with an issue ticket there.

File: ChangeLog [hunk 2]
```diff
@@ -1,4 +1,24 @@
...
+	* src/LinReg.cpp: Boolean return for MCMC function and
+	setting MCMC repeats separately.
+	* src/LinReg.cpp: Idem.
+	* inst/include/LinReg.h: Idem.
+	* inst/include/LinReg_LA.h: Idem.
+
+	* src/RcppExports.cpp: Regenerated.
+	* R/RcppExports.R: Idem.
+
+2017-08-04  Adam M. Johansen  <adam.johansen@gmail.com>
```
Comment (eddelbuettel) [inline ChangeLog]: Thanks for cleaning that up.

File: R/RcppExports.R (hunk 1 above)
Comment (LeahPrice) [inline R/RcppExports.R]: Thanks for picking up on that. I've fixed it up now.

Comment (eddelbuettel) [inline R/RcppExports.R]: Thanks.

**Comment by adamjohansen** [comment]
Thanks, both -- looks good.

**Pull request merged**